### DISCUSSION TITLE
How can I disable the 'format code on save"

### DISCUSSION THREAD

**Discussion by x2c3z4** [adoyle-h/one.nvim#6] [Q&A] [answered]
When saving buffer to disk, the code is formatted, how can I close it.

**Comment by adoyle-h** [comment, marked as answer]
## Disable formatting for temporary

If you want to disable it for temporary, use `:ToggleLSP` to disable LSP (But it will be enabled after refreshing buffer) for current buffer. And `:ToggleNullLSP` to disable NullLS formatting source for all buffers.

## Disable formatting on-save for permanent

If you want to only disable formatting on-save,

```lua
require('one').setup {
	config = {
		lsp = {
			format = {
				_ = { on_save = true },

				c = { on_save = false },

				lua = {
					on_save = true,
				},
			},
		},
	},
}
```

For above config, `c` filetype buffers will disable formatting on-save, while other filetypes enabled.

The `c` and `lua` key are filetype. You can change it as your demand.
the "_" will match all filetypes. Please read [this link](https://github.com/adoyle-h/lsp-format.nvim#global-format-options).

You can press `<M-=>` to format whole file, or press `<M-=>` to format text in selection. Read [doc/usage/keymaps.md](https://github.com/adoyle-h/one.nvim/blob/master/doc/usage/keymaps.md#format).

This feature are implemented in [adoyle-h/lsp-format.nvim](https://github.com/adoyle-h/lsp-format.nvim) (a fork from [lukas-reineke/lsp-format.nvim](https://github.com/lukas-reineke/lsp-format.nvim)) and [plugins/lsp/format](https://github.com/adoyle-h/one.nvim/blob/master/lua/one/plugins/lsp/format.lua).

[The `config.lsp.format` is passed to `lspFormat.setup`](https://github.com/adoyle-h/one.nvim/blob/0053166018dac893acdeb1601a9ea55877af5664/lua/one/plugins/lsp/format.lua#L31).

## Disable formatting for permanent

If you want to completely disable it for permanent,

```lua
require('one').setup {
	config = {
		lsp = {
			format = {
				lua = {
					exclude = { -- exclude is a table of LSP servers that should not format the buffer whose filetype matching "lua".
						'sumneko_lua',
						'null-ls',
					},
				},
			},
		},
	}
}
```

Read `:h lsp-format-exclude` for `exclude` option.

**Reply by adoyle-h** [reply]
If this answer solves your problem, please click "mark as answer" button.

**Reply by x2c3z4** [reply]
I want to disable the .cc/.cpp/.h format, how can I do?
Right?
```
			format = {
				cc = {
					exclude = { -- exclude is a table of LSP servers that should not format the buffer whose filetype matching "lua".
						'clang-format',
						'clangd',
					},
				},
				h = {
					exclude = { -- exclude is a table of LSP servers that should not format the buffer whose filetype matching "lua".
						'clang-format',
						'clangd',
					},
				},
				cpp = {
					exclude = { -- exclude is a table of LSP servers that should not format the buffer whose filetype matching "lua".
						'clang-format',
						'clangd',
					},
				},
			},
		},
	},
```

**Reply by x2c3z4** [reply]
I think I want to disable all auto format. And If I need to format my code, using 'gF' to format the whole buffer.

Btw, how can I format just a code block?

**Reply by adoyle-h** [reply]
Open your .cc, .h, .cpp files, the filetype shows on the right of status line.
<img width="169" alt="CleanShot 2022-12-05 at 17 03 57@2x" src="https://user-images.githubusercontent.com/1998490/205597079-77c5431e-74a2-48b7-9b16-1ef0ff4401d0.png">

Or, you can invoke `:set ft` to get filetype.

**Reply by adoyle-h** [reply]
> I think I want to disable all auto format. And If I need to format my code, using 'gF' to format the whole buffer. Btw, how can I format just a code block?

Sorry, it is not supported for now. I will consider to implement it later.

**Reply by x2c3z4** [reply]
ok, thanks.

**Reply by adoyle-h** [reply]
The formatting is based on LSP. For formatting in range, we should use the [`textDocument/rangeFormatting`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocumentClientCapabilities) capability and pass [`DocumentRangeFormattingParams`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#documentRangeFormattingParams) to LSP.
But I find only a very few LSPs have implemented `textDocument/rangeFormatting` capability. Most LSPs only implemented `textDocument/formatting`.
So I think it is not ready to implement this feature in near future.

**Reply by x2c3z4** [reply]
VScode support it, so may be has other method to do it, or just support some languages, like c/cpp.

**Reply by adoyle-h** [reply]
[VS Code has default formatters for JavaScript, TypeScript, JSON, and HTML](https://code.visualstudio.com/Docs/editor/codebasics#_formatting). For example, [typescript](https://github.com/microsoft/vscode/blob/main/extensions/typescript-language-features/src/languageFeatures/formatting.ts) and [HTML](https://github.com/microsoft/vscode/blob/main/extensions/html-language-features/server/src/modes/formatting.ts), they are VSCode extensions, but not LSP.

I find that clangd supports rangeFormatting. I will support it soon.

**Reply by adoyle-h** [reply]
Now the formatting in range is supported. Please update one.nvim to latest commit, and update lsp-format.nvim to my fork ([adoyle-h/lsp-format.nvim](https://github.com/adoyle-h/lsp-format.nvim)).

I have updated the answer and one.nvim document.

Notice:

1. the keymap for formatting is changed from `gF` to `<M-=>`
2. Only few LSPs support formatting in range. You can try `clangd` to format c/c++.

  <img width="781" alt="CleanShot 2022-12-07 at 14 12 38@2x" src="https://user-images.githubusercontent.com/1998490/206102484-9ce1c90c-0cb7-49d8-93fc-64e2f9365bde.png">

**Reply by x2c3z4** [reply]
Great job!

**Reply by x2c3z4** [reply]
I just try the latest version.
The 'format' doesn't work, press 'M-=' and nothing happens.

**Reply by adoyle-h** [reply]
1. `<M-=>` means holding `alt` key (it is ⌥ in MacOS) and press `=`, is it right?
2. Check your lsp-format.nvim is [adoyle-h/lsp-format.nvim](https://github.com/adoyle-h/lsp-format.nvim) instead of [lukas-reineke/lsp-format.nvim](https://github.com/lukas-reineke/lsp-format.nvim) in your plugin directory.
  For vim-plug, plugin directory is `~/.local/share/nvim/plugins`.
  For packer, plugin directory is `~/.local/share/nvim/pack/packer/start/`.
3. Check your LSP installed. Press `<M-m>` to open mason window.
4. Open your file, and invoke `:LspInfo` to check the LSP is attached to buffer.
5. Press `<space>n` to check whether there are any error messages or not.

**Reply by x2c3z4** [reply]
I have added this in the init.lua, because if don't add this, the code will be auto formatted.
So I think we should firstly disable the auto save, then support format code manually.

```
        config = {
                lsp = {
                        format = {
                                c = {
                                        exclude = { -- exclude is a table of LSP servers that should not format the buffer whose filetype matching "lua".
                                                'clang-format',
                                                'clangd',
                                        },
                                },
                                cc = {
                                        exclude = { -- exclude is a table of LSP servers that should not format the buffer whose filetype matching "lua".
                                                'clang-format',
                                                'clangd',
                                        },
                                },
                                h = {
                                        exclude = { -- exclude is a table of LSP servers that should not format the buffer whose filetype matching "lua".
                                                'clang-format',
                                                'clangd',
                                        },
                                },
                                cpp = {
                                        exclude = { -- exclude is a table of LSP servers that should not format the buffer whose filetype matching "lua".
                                                'clang-format',
                                                'clangd',
                                        },
                                },
                        },
```

**Reply by adoyle-h** [reply]
If you use `exclude`, it will disable format completely. You should use `on_save = false`. Read the `Disable formatting on-save for permanent` section in [this answer](https://github.com/adoyle-h/one.nvim/discussions/6#discussioncomment-4309726).

**Reply by x2c3z4** [reply]
Thanks, it works.
